node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java

```diff
@@ -809,13 +814,19 @@ else if (!node.type().isHost()) { // removing a child node
         // perform operation while holding locks
         List<Node> resultingNodes = new ArrayList<>();
         try (Mutex lock = lockUnallocated()) {
-            for (Node node : unallocatedNodes)
-                resultingNodes.add(action.apply(node, lock));
+            for (Node node : unallocatedNodes) {
+                Optional<Node> currentNode = db.readNode(node.hostname()); // Re-read while holding lock
```

Comment:
This is an improvement.

I'm a bit sceptical of the method API: Usually, some condition has led to the caller wanting to set this node to dirty.  I would guess that whole condition needs to be reevaluated under the unallocated lock?  More than just the presence of the node may have changed between testing the condition and coming to this point where the lock is held.

Reply:
Agree, the API is not ideal.